workitem.go

```diff
@@ -1,177 +1,187 @@
+	result := app.WorkItem{
+		ID:      strconv.FormatUint(workItem.ID, 10),
+		Name:    workItem.Name,
+		Type:    workItem.Type,
+		Version: workItem.Version,
+		Fields:  map[string]interface{}{}}
+
+	for name, field := range wiType.Fields {
+		var err error
+		result.Fields[name], err = field.ConvertFromModel(name, workItem.Fields[name])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+// Create runs the create action.
+func (c *WorkitemController) Create(ctx *app.CreateWorkitemContext) error {
+	wiType, err := loadTypeFromDB(ctx.Payload.Type)
+	if err != nil {
+		return ctx.BadRequest(goa.ErrBadRequest(err.Error()))
+	}
+	wi := models.WorkItem{
+		Name:   ctx.Payload.Name,
+		Type:   ctx.Payload.Type,
+		Fields: models.Fields{},
+	}
+
+	for fieldName, fieldDef := range wiType.Fields {
+		fieldValue := ctx.Payload.Fields[fieldName]
+		var err error
+		wi.Fields[fieldName], err = fieldDef.ConvertToModel(fieldName, fieldValue)
+		if err != nil {
+			return ctx.BadRequest(goa.ErrBadRequest("Could not parse id: %s", err.Error()))
+		}
+	}
+	tx := c.db.Begin()
+
+	if tx.Create(&wi).Error != nil {
+		tx.Rollback()
+		return ctx.InternalServerError()
+	}
+	log.Printf("created item %v\n", wi)
+	result, err := convertFromModel(*wiType, wi)
+	if err != nil {
+		tx.Rollback()
+		return ctx.InternalServerError()
+	}
+	tx.Commit()
+	return ctx.OK(result)
+}
+
+// Delete runs the delete action.
+func (c *WorkitemController) Delete(ctx *app.DeleteWorkitemContext) error {
+	var workItem models.WorkItem = models.WorkItem{}
+	id, err := strconv.ParseUint(ctx.ID, 10, 64)
+	if err != nil {
+		return ctx.BadRequest(goa.ErrBadRequest("Could not parse id: %s", err.Error()))
+	}
+	tx := c.db.Begin()
+
+	if tx.First(&workItem, id).RecordNotFound() {
+		log.Print("not found, res=%v", workItem)
+		tx.Rollback()
+		return ctx.NotFound()
+	}
+	if tx.Delete(workItem).Error != nil {
+		tx.Rollback()
+		return ctx.InternalServerError()
+	}
+
+	tx.Commit()
+	return ctx.OK([]byte{})
+}
+
+// Update runs the update action.
+func (c *WorkitemController) Update(ctx *app.UpdateWorkitemContext) error {
+	res := models.WorkItem{}
+	id, err := strconv.ParseUint(ctx.Payload.ID, 10, 64)
+	if err != nil {
+		return ctx.BadRequest(goa.ErrBadRequest("Could not parse id: %s", err.Error()))
+	}
+
+	log.Printf("looking for id %d", id)
+	tx := c.db.Begin()
+	if tx.First(&res, id).RecordNotFound() {
+		tx.Rollback()
+		log.Printf("not found, res=%v", res)
+		return ctx.NotFound()
+	}
+	if res.Version != ctx.Payload.Version {
+		tx.Rollback()
+		return ctx.BadRequest(goa.ErrBadRequest("version conflict: expected %d but got %d", res.Version, ctx.Payload.Version))
+	}
+
+	wiType, err := loadTypeFromDB(res.Type)
+	if err != nil {
+		return ctx.BadRequest(goa.ErrBadRequest(err.Error()))
+	}
+
+	wi := models.WorkItem{
+		ID: id,
+		Name:    ctx.Payload.Name,
+		Type:    ctx.Payload.Type,

```

Comment:
The missing tx.Rollback() is unrelated to the issue. You're technically not changing anything in the db yet at this point so there is nothing to Rollback.

The problem is; You load up the WorkItem from the database into res. Then load the WorkItemType based on res.Type. So all good so far, since Type was accepted during Create, we have a Type. Then you create a new WorkItem with the values from the Request and store it on the same id. The Request.Type on the other hand, is/can be nil/"". Now you effectively stored a WorkItem with no associated Type and Get/Show fail.